# 🏡 Min-Max Normalization Workshop
## Team Name: Group 1
## Team Members: Zhuoran Zhang, Ce Chen
---

## ❗ Why We Normalize: The Problem with Raw Feature Scales

In housing data, features like `Price` and `Lot_Size` can have values in the hundreds of thousands, while others like `Num_Bedrooms` range from 1 to 5. This creates problems when we use algorithms that depend on numeric magnitudes.

---

### ⚠️ What Goes Wrong Without Normalization

---

### 1. 🧭 K-Nearest Neighbors (KNN)

KNN uses the **Euclidean distance** formula:

$$
d = \sqrt{(x_1 - x_2)^2 + (y_1 - y_2)^2 + \cdots}
$$

**Example:**

- $ \text{Price}_1 = 650{,}000, \quad \text{Price}_2 = 250{,}000 $
- $ \text{Bedrooms}_1 = 3, \quad \text{Bedrooms}_2 = 2 $

Now compute squared differences:

$$
(\text{Price}_1 - \text{Price}_2)^2 = (650{,}000 - 250{,}000)^2 = (400{,}000)^2 = 1.6 \times 10^{11}
$$
$$
(\text{Bedrooms}_1 - \text{Bedrooms}_2)^2 = (3 - 2)^2 = 1
$$

➡️ **Price dominates the distance calculation**, making smaller features like `Bedrooms` irrelevant.

---

### 2. 📉 Linear Regression

Linear regression estimates:

$$
y = \beta_1 \cdot \text{Price} + \beta_2 \cdot \text{Bedrooms} + \beta_3 \cdot \text{Lot\_Size} + \epsilon
$$

If `Price` has very large values:
- Gradient updates for $ \beta_1 $ will be **much larger**
- Gradient updates for $ \beta_2 $ (Bedrooms) will be **very small**

➡️ The model overfits high-magnitude features like `Price`.

---

### 3. 🧠 Neural Networks

A single neuron computes:

$$
z = w_1 \cdot \text{Price} + w_2 \cdot \text{Bedrooms} + w_3 \cdot \text{Lot\_Size}
$$

If:

- $ \text{Price} = 650{,}000 $
- $ \text{Bedrooms} = 3 $
- $ \text{Lot\_Size} = 8{,}000 $

Then:

$$
z \approx w_1 \cdot 650{,}000 + w_2 \cdot 3 + w_3 \cdot 8{,}000
$$

➡️ Even with equal weights, `Price` contributes **most of the activation**, making it difficult for the network to learn from other features.

---

### ✅ Solution: Min-Max Normalization

We apply the transformation:

$$
x_{\text{normalized}} = \frac{x - x_{\text{min}}}{x_{\text{max}} - x_{\text{min}}}
$$

This scales all features to a common range (typically $[0, 1]$).

| Feature      | Raw Value | Min     | Max     | Normalized Value |
|--------------|-----------|---------|---------|------------------|
| Price        | 650,000   | 250,000 | 800,000 | 0.72             |
| Bedrooms     | 3         | 1       | 5       | 0.50             |
| Lot_Size     | 8,000     | 3,000   | 10,000  | 0.714            |

➡️ Now, **each feature contributes fairly** to model training or distance comparisons.

---

## 📌 Use Case: Housing Data
We are normalizing features from a real estate dataset to prepare it for machine learning analysis.

In [1]:
# 🔢 Load and display dataset
import pandas as pd
df = pd.read_csv('./data/housing_data.csv')
df.head()

,House_ID,Price,Area_sqft,Num_Bedrooms,Num_Bathrooms,Year_Built,Lot_Size
0,H100000,574507,1462,3,3,2002,4878
1,H100001,479260,1727,2,2,1979,4943
2,H100002,597153,1403,5,2,1952,5595
3,H100003,728454,1646,5,2,1992,9305
4,H100004,464876,853,1,1,1956,7407


### 🔎 Step 1 — Implement Min-Max Normalization on the Housing Dataset

In [2]:
# ✍️ Implement Min-Max Normalization manually here (no sklearn/numpy)
# Normalize: Price, Area_sqft, Num_Bedrooms, Num_Bathrooms, Lot_Size
def min_max_normalize(column):
    min_val = column.min()
    max_val = column.max()
    return (column - min_val) / (max_val - min_val)

df['Price_norm'] = min_max_normalize(df['Price'])
df['Area_sqft_norm'] = min_max_normalize(df['Area_sqft'])
df['Num_Bedrooms_norm'] = min_max_normalize(df['Num_Bedrooms'])
df['Num_Bathrooms_norm'] = min_max_normalize(df['Num_Bathrooms'])
df['Lot_Size_norm'] = min_max_normalize(df['Lot_Size'])
df.head()

,House_ID,Price,Area_sqft,Num_Bedrooms,Num_Bathrooms,Year_Built,Lot_Size,Price_norm,Area_sqft_norm,Num_Bedrooms_norm,Num_Bathrooms_norm,Lot_Size_norm
0,H100000,574507,1462,3,3,2002,4878,0.485226,0.315789,0.50,1.0,0.320814
1,H100001,479260,1727,2,2,1979,4943,0.387827,0.394588,0.25,0.5,0.326191
2,H100002,597153,1403,5,2,1952,5595,0.508384,0.298246,1.00,0.5,0.380129
3,H100003,728454,1646,5,2,1992,9305,0.642651,0.370503,1.00,0.5,0.687045
4,H100004,464876,853,1,1,1956,7407,0.373119,0.134701,0.00,0.0,0.530030


### 🔎 Talking Points #1 — Min-Max Normalization

The original features had very different scales (price and area were large numbers, while bedrooms and bathrooms were small integers).

After Min-Max normalization, all features were scaled to the range 0–1, so each feature contributes more equally to the analysis.

This is important for distance-based models like KNN because large-value features would otherwise dominate the distance calculation.


## 🧩 Challenge Extension: After Normalization, Which Features Matter Most?

You’ve normalized the housing features so they share a common scale.  
Now comes a common next step in ML workflows:

> **How do we identify the most important directions (principal components) in the data—and how might these relate to a target variable like `Price`?**

This introduces **Principal Component Analysis (PCA)**.

---

## 📚 PCA Theory (Conceptual)

### What PCA *is*
PCA is an **unsupervised** dimensionality reduction technique that:
- Finds **new axes** (principal components) that are **linear combinations** of your original features.
- Orders these axes so that:
  - **PC1** captures the **most variance** in the feature space,
  - **PC2** captures the next most variance, and so on,
  - Each PC is **orthogonal** (uncorrelated) with the previous ones.

### What PCA is *not*
PCA does **not** directly find features that “impact the target variable” because it does not use the target in its optimization.

However, you *can*:
- Compute PCs from the feature matrix **X** (after normalization),
- Then measure how PCs relate to the target **y** (e.g., correlation with `Price`, or a simple regression on PCs),
- Interpret which original features contribute most to PCs that are most related to **y**.

---

## 🧠 The Math (high level)
Given a centered feature matrix \(X\) (often standardized/normalized first):

1. Compute covariance matrix:
$$
\Sigma = \frac{1}{n-1}X^\top X
$$

2. Find eigenvectors (principal directions) and eigenvalues (variance captured):
$$
\Sigma v_i = \lambda_i v_i
$$

- $ v_i $ are **principal component directions** (loadings)
- $ \lambda_i $ are the **variance explained** by each component

---

## ✅ Why Normalize Before PCA?
PCA is sensitive to scale. Without normalization/standardization:
- A large-scale feature (e.g., `Price`) can dominate variance
- PCs will reflect units rather than structure

---

## 🎯 Student Challenge
Using the **housing dataset**:

1. Apply PCA to the normalized feature matrix \(X\) (exclude ID columns and the target).
2. Determine how many components are needed to explain **≥ 90%** of the variance.
3. Identify which original features contribute most to:
   - **PC1** and **PC2**, and
   - the **PC most correlated with the target** (`Price`).
4. Write a short interpretation:
   - “What does PC1 represent in housing terms?”
   - “Do the PCs that explain the most variance also relate most strongly to `Price`?”



### 🔗 How to Integrate This With Your Step 1 Normalization

- If you created normalized columns (e.g., `Area_sqft_norm`), use those in `candidate_features`.
- If you normalized in-place (overwriting original columns), you can use the original names.
- PCA should **not** include:
  - `House_ID` (identifier)
  - non-numeric categorical columns (unless encoded appropriately)
- Decide intentionally whether to include `Year_Built`:
  - It’s numeric, but it may behave differently than size/price-related features.

---

### ✅ Deliverable for the Challenge
Add a Markdown cell answering:

1. How many PCs explain at least **90%** variance?
2. Which features contribute most to **PC1** and **PC2**?
3. Which PC is most correlated with `Price`?
4. In plain language: what do you think PC1 represents?


In [3]:
# --- PCA Scaffolding (after you have a normalized DataFrame) ---
# Assumptions:
# 1) You already loaded housing_data.csv into df
# 2) You created normalized columns OR a normalized DataFrame (e.g., df_norm)
#
# If you normalized in-place, you can build df_norm by selecting only the normalized feature columns.

import pandas as pd
import numpy as np

from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler


# ---- Step A: Choose target and feature columns ----
target = "Price"

feature_cols = [
    "Area_sqft_norm",
    "Num_Bedrooms_norm",
    "Num_Bathrooms_norm",
    "Lot_Size_norm"
]

X = df[feature_cols]
y = df[target]


# ---- Step B: Ensure scaling appropriate for PCA ----
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


# ---- Step C: Fit PCA ----
pca = PCA()
X_pca = pca.fit_transform(X_scaled)


# ---- Step D: Variance explained ----
print("Explained Variance Ratio:")
for i, var in enumerate(pca.explained_variance_ratio_):
    print(f"PC{i+1}: {var:.4f}")


# ---- Step E: Loadings (feature contributions to PCs) ----
loadings = pd.DataFrame(
    pca.components_.T,
    columns=[f"PC{i+1}" for i in range(len(feature_cols))],
    index=feature_cols
)

print("\nFeature Loadings:")
print(loadings)


# ---- Step F: Relate PCs to the target (simple correlation) ----
pc1 = X_pca[:, 0]
correlation = pd.Series(pc1).corr(y)

print(f"\nCorrelation between PC1 and Price: {correlation:.4f}")


Explained Variance Ratio:
PC1: 0.2654
PC2: 0.2523
PC3: 0.2424
PC4: 0.2399

Feature Loadings:
                         PC1       PC2       PC3       PC4
Area_sqft_norm      0.579685 -0.052234  0.768064 -0.267049
Num_Bedrooms_norm   0.602727  0.246080 -0.181941  0.736928
Num_Bathrooms_norm  0.097440  0.890329 -0.157600 -0.415911
Lot_Size_norm      -0.539621  0.379513  0.593412  0.461130

Correlation between PC1 and Price: -0.0225


### 🔎 Talking Points #2 — PCA Results

The explained variance is spread across all principal components, meaning no single feature dominates the dataset.

The loadings show PC1 is mainly influenced by area and number of bedrooms, while PC2 is influenced by number of bathrooms.

The correlation between PC1 and price is very low (−0.0225), suggesting housing price is not strongly determined by these physical features alone.